2.5-huffman.ipynb

```diff
@@ -1,25 +1,207 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To construct a TapTree with the Huffman algorithm, we simply assign a relative probability weight to each leaf. \n",
+    "\n",
+    "A higher `leaf weight` represents:\n",
+    "\n",
+    "* A higher likelihood of execution\n",
+    "* A position which is closer to the root (if possible)\n",
+    "* A smaller inclusion proof"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 1: Huffman Algorithm\n",
+    "\n",
+    "The Huffman algorithm to construct a TapTree structure can be described as follows:\n",
+    "\n",
+    "* Place all leafs in a node queue and sort by weight*\n",
+    "* While length ( node queue ) > 1 :\n",
+    "    * Pop the two elements with the highest weights\n",
+    "    * Generate parent node from popped elements\n",
+    "    * Add parent node to queue and resort\n",
+    "* Root node is represented by final queue element"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/huffman_intro0.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ">\\* If more than 2 leafs or nodes share the same weight during construction with the Huffman algorithm, the selection of the two nodes to generate the next parent becomes ambiguous. In our implementation, we first sort our node queue by `weight`, then by `tagged hash` value. This way, the sorting of the node queue is always unambiguous given a distinct set of leafs."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.5.1: Construct a TapTree with the Huffman algorithm\n",
+    "\n",
+    "* We manually construct the TapTree from a set of 5 pay-to-pubkey TapLeafs with the same weights as shown in the image above."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "internal_pubkey = ECPubKey()\n",
+    "internal_pubkey.set(bytes.fromhex('03af455f4989d122e9185f8c351dbaecd13adca3eef8a9d38ef8ffed6867e342e3'))\n",
+    "\n",
+    "# Derive pay-to-pubkey tapleafs\n",
+    "privkeyA = ECKey().generate()\n",
+    "privkeyB = ECKey().generate()\n",
+    "privkeyC = ECKey().generate()\n",
+    "privkeyD = ECKey().generate()\n",
+    "privkeyE = ECKey().generate()\n",
+    "\n",
+    "pubkeyA = privkeyA.get_pubkey()\n",
+    "pubkeyB = privkeyB.get_pubkey()\n",
+    "pubkeyC = privkeyC.get_pubkey()\n",
+    "pubkeyD = privkeyD.get_pubkey()\n",
+    "pubkeyE = privkeyD.get_pubkey()\n",
+    "\n",
+    "tapleafA = TapLeaf()\n",
+    "tapleafB = TapLeaf()\n",
+    "tapleafC = TapLeaf()\n",
+    "tapleafD = TapLeaf()\n",
+    "tapleafE = TapLeaf()\n",
+    "\n",
+    "tapleafA.construct_pk(pubkeyA)\n",
```

Comment:
Can we make `construct_pk()` return `self` and make the construction look like:

```
tapleafA = Tapleaf().construct_pk(pubkeyA)
```

I think that's more compact and easier to understand.